KD_Lib/MeanTeacher/MeanTeacher.py

```diff
@@ -0,0 +1,83 @@
+from torch import nn
+from KD_Lib.common import BaseClass
+
+
+class MeanTeacher(BaseClass):
+    """
+    Implementation of Knowledge distillation using a mean teacher from the
+    paper "Mean teachers are better role models:Weight-averaged consistency
+    targets improvesemi-supervised deep learning results"
+    https://arxiv.org/abs/1703.01780
+
+    :param teacher_model (torch.nn.Module): Teacher model
+    :param student_model (torch.nn.Module): Student model
+    :param train_loader (torch.utils.data.DataLoader): Dataloader for training
+    :param val_loader (torch.utils.data.DataLoader): Dataloader for validation
+    :param optimizer_teacher (torch.optim.*): Optimizer for training teacher
+    :param optimizer_student (torch.optim.*): Optimizer for training student
+
+    :param loss (str): Consistency criterion for loss
+    :param class_loss (torch.nn.Module): Class Criterion for loss
+    :param res_loss (torch.nn.Module): Residual Logit Criterion for loss
+
+    :param temp (float): Temperature parameter for distillation
+    :param distil_weight (float): Weight paramter for distillation loss
+    :param device (str): Device for training; 'cpu' for cpu and 'cuda' for gpu
+    :param log (bool): True if logging required
+    :param logdir (str): Directory for storing logs
+    """
+    def __init__(self, teacher_model, student_model, train_loader, val_loader,
+                 optimizer_teacher, optimizer_student, loss='MSE',
+                 class_loss=nn.MSELoss(), res_loss=nn.MSELoss(), temp=20.0,
+                 distil_weight=0.5, device='cpu', log=False,
+                 logdir='./Experiments'):
+        super(MeanTeacher, self).__init__(teacher_model,
+                                          student_model,
+                                          train_loader,
+                                          val_loader,
+                                          optimizer_teacher,
+                                          optimizer_student,
+                                          loss,
+                                          temp,
+                                          distil_weight,
+                                          device,
+                                          log,
+                                          logdir)
+        if self.loss.upper() == 'MSE':
+            self.loss_fn = nn.MSELoss()
+
+        elif self.loss.upper() == 'KL':
+            self.loss_fn = nn.KLDivLoss()
+        else:
+            assert False, "Invalid loss"
```

Comment:
Better to raise an exception right?

Reply:
Yeah that makes sense. On the larger scale, `loss_fn` is used by many implementations, TAKD, self_training, virtual_training, orignal_paper.
Should I also move this to `BaseClass`?

Reply:
Yea you can move this to BaseClass